src/alg/rsa.rs

```diff
@@ -0,0 +1,292 @@
+        let e = BigUint::from_bytes_be(e);
+        match RSAPublicKey::new(n, e) {
+            Err(e) => Err(RsaError::InvalidKey(e).into()),
+            Ok(key) => Ok(RsaVerifyingKey(key)),
+        }
+    }
+}
+
+/// An RSA signing key.
+#[derive(Debug)]
+pub struct RsaSigningKey(RSAPrivateKey);
+
+impl AsRef<RSAPrivateKey> for RsaSigningKey {
+    fn as_ref(&self) -> &RSAPrivateKey {
+        &self.0
+    }
+}
+
+impl RsaSigningKey {
+    /// Create a signing key from a DER-encoded set of the parameters.
+    pub fn from_der(der: &[u8]) -> anyhow::Result<RsaSigningKey> {
+        match RSAPrivateKey::from_pkcs8(&der) {
+            Err(e) => Err(RsaError::InvalidKey(e).into()),
+            Ok(key) => {
+                key.validate().map_err(RsaError::InvalidKey)?;
+                Ok(RsaSigningKey(key))
+            }
+        }
+    }
+
+    /// Convert a signing key to a verification key.
+    pub fn to_verifying_key(&self) -> RsaVerifyingKey {
+        RsaVerifyingKey(RSAPublicKey::from(&self.0))
+    }
+}
+
+/// Integrity algorithm using digital signatures on RSA-PKCS1v1.5 and SHA-256.
+///
+/// The name of the algorithm is specified as `RS256` as per the [IANA registry].
+///
+/// *This type is available if the crate is built with the `rsa` feature.*
+///
+/// [IANA registry]: https://www.iana.org/assignments/jose/jose.xhtml
+#[derive(Debug)]
+pub struct Rsa {
+    hash_alg: Hash,
+    padding_alg: Padding,
+}
+
+impl Rsa {
+    /// Create an instance using a specific hash function and padding
+    pub fn new(hash_alg: Hash, padding_alg: Padding) -> Self {
+        Rsa {
+            hash_alg,
+            padding_alg,
+        }
+    }
+}
+
+pub trait RsaVariant {
+    fn rsa() -> Rsa;
+}
+
+impl<T: RsaVariant> Algorithm for T {
+    type SigningKey = RsaSigningKey;
+    type VerifyingKey = RsaVerifyingKey;
+    type Signature = Signature;
+
+    fn name(&self) -> Cow<'static, str> {
+        Self::rsa().name()
+    }
+
+    fn sign(&self, signing_key: &Self::SigningKey, message: &[u8]) -> Self::Signature {
+        Self::rsa().sign(signing_key, message)
+    }
+
+    fn verify_signature(
+        &self,
+        signature: &Self::Signature,
+        verifying_key: &Self::VerifyingKey,
+        message: &[u8],
+    ) -> bool {
+        Self::rsa().verify_signature(signature, verifying_key, message)
+    }
+}
+
+impl Rsa {
+    fn hash(&self, message: &[u8]) -> Vec<u8> {
+        match self.hash_alg {
+            Hash::SHA2_256 => Sha256::digest(message).to_vec(),
+            Hash::SHA2_384 => Sha384::digest(message).to_vec(),
+            Hash::SHA2_512 => Sha512::digest(message).to_vec(),
+            _ => unreachable!(),
+        }
+    }
+
+    fn padding_scheme(&self) -> PaddingScheme {
+        match self.padding_alg {
+            Padding::Pkcs1v15 => PaddingScheme::new_pkcs1v15_sign(Some(self.hash_alg)),
+            Padding::Pss => {
+                let rng = rand_core::OsRng {};
```

Comment:
> This would allow varying it according to user's requirements

The RNG is used for two things:

- To generate keys. If a weak RNG is used, this has catastrophic security implications.
- To generate RSA padding and p256 nonces. If a weak RNG is used, this has catastrophic security implications as well.

In both cases, allowing users to provide their RNG is dangerous. It complicates the user-facing API. It makes the internal code more complicated and more bug-prone.

While the best option in virtually all cases is simply to use `OsRng`.

Would you possibly consider using it automatically instead of relying on users providing their own RNG?

Reply:
Also, an RNG is not required for verification.

So, requiring one in the constructor would be a bit weird when the intent is to very tokens.